### PR TITLE
Fix broken tag query message replay logic

### LINKED ISSUE #385: CurrentEventsByTag returns a maximum of 1,000 events
I know that there are 19,000 events in the log for the given tag. The query is only giving me the first 1,000.

Configuration:

```
.WithAzurePersistence(
    azureStorageConnectionString,
    true,
    containerName,
    tableName,
    builder =>
    {
        builder.AddWriteEventAdapter<TaggingEventAdapter>(
            "projection-tagger",
            new[] { typeof(IHasEntraTenantId) }
        );
    }
)
```

Query:

```
private void QueryEvents()
{
    var materializer = Context.System.Materializer();
    var projectionActorRef = Self;

    currentEventsByTagQuery
        .CurrentEventsByTag(tag, Offset.NoOffset())
        .Select(env => env.Event)
        .RunWith(
            Sink.ActorRef<object>(
                projectionActorRef,
                new ProjectionRebuildNoMoreEvents { EntraTenantId = entraTenantId },
                (ex) =>
                    new FailedEndOfStream { EntraTenantId = entraTenantId, Reason = ex.Message }
            ),
            materializer
        );
}
```

When I swap to the Akka.Persistence.Sql plugin I get all 19,000 events.

Thanks, 

Kyle Reed

### PR DESCRIPTION
**Pull request by Arkatufus**
Fixes #385

## Changes
* Change `AzureTableStorageReadJournal.CurrentEventsByTag` `toOffset` to `DateTime.UtcNow.Ticks` to enable query limiting
* Change how `AzureTableStorageReadJournal.ReplayTaggedMessagesAsync()` signals if it reaches the end of records
* Limit the number of record fetch to limit network bandwidth waste
* Fix `CurrentEventsByTagPublisher` logic to determine completion from the read journal return status.

### CODE REVIEW

File: src/Akka.Persistence.Azure/Query/Publishers/CurrentEventsByTagPublisher.cs [hunk 1]
```diff
@@ -41,14 +41,14 @@ protected override void ReceiveInitialRequest()
...
         {
+            if (completed)
+                _completed = true;
+            
             Buffer.DeliverBuffer(TotalDemand);
 
-            if (highestSequenceNr < ToOffset)
-                _toOffset = highestSequenceNr;
-
-            if (Buffer.IsEmpty && CurrentOffset > ToOffset)
```
Comment (Arkatufus) [inline src/Akka.Persistence.Azure/Query/Publishers/CurrentEventsByTagPublisher.cs:51]: This code is a vestigial code from SQLite implementation. There is no way for the publisher to know if it received all available tagged records by highest sequence number alone. 
Also, `_toOffset = highestSequence` only works if the database only returns a single page result (the actual cause of the bug).

File: src/Akka.Persistence.Azure/Query/Messages.cs [hunk 2]
```diff
@@ -172,6 +172,16 @@ public ReplayTaggedMessages(long fromOffset, long toOffset, long max, string tag
...
 
+    public sealed class ReplayTaggedMessageSuccess
+    {
+        public ReplayTaggedMessageSuccess(bool completed)
+        {
+            Completed = completed;
+        }
+
+        public bool Completed { get; }
+    }
```
Comment (Arkatufus) [inline src/Akka.Persistence.Azure/Query/Messages.cs:183]: New message to signal that the database have returned all available rows and the operation has been completed.

File: src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs [hunk 3]
```diff
@@ -761,16 +755,17 @@ private async Task<long> ReplayTaggedMessagesAsync(ReplayTaggedMessages replay,
                         counter++;
                     }
 
-                    maxOrderingId = Math.Max(maxOrderingId, entry.UtcTicks);
+                    currentOrderingId = Math.Max(currentOrderingId, entry.UtcTicks);
+                    
+                    if (counter >= replay.Max)
+                        return new ReplayTaggedMessageSuccess(false);
```
Comment (Arkatufus) [inline src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs:758]: Reached the maximum records that can fit inside the publisher buffer, signal that we have not completed retrieving all tagged records.

File: src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs [hunk 4]
```diff
@@ -761,16 +755,17 @@ private async Task<long> ReplayTaggedMessagesAsync(ReplayTaggedMessages replay,
...
                 }
 
                 if (counter >= replay.Max)
-                {
-                    break;
-                }
+                    return new ReplayTaggedMessageSuccess(false);
```
Comment (Arkatufus) [inline src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs:762]: Reached the maximum records that can fit inside the publisher buffer, signal that we have not completed retrieving all tagged records.

File: src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs [hunk 5]
```diff
@@ -761,16 +755,17 @@ private async Task<long> ReplayTaggedMessagesAsync(ReplayTaggedMessages replay,
...
             }
 
-            return maxOrderingId;
+            return new ReplayTaggedMessageSuccess(true);
```
Comment (Arkatufus) [inline src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs:765]: Records returned by all page is less than the maximum requested, we have reached the last row from the query.

**Review by Arkatufus — COMMENTED** [review]
Self review

File: src/Akka.Persistence.Azure/Query/Publishers/CurrentEventsByTagPublisher.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/Akka.Persistence.Azure/Query/Publishers/CurrentEventsByTagPublisher.cs:51]: Sigh, nice catch.

File: src/Akka.Persistence.Azure.Tests/Query/AzureTableCurrentEventsByTagSpec.cs [hunk 6]
```diff
@@ -32,5 +39,67 @@ public AzureTableCurrentEventsByTagSpec(ITestOutputHelper output)
...
+            if (ReadJournal is not ICurrentEventsByTagQuery queries)
+                throw IsTypeException.ForMismatchedType(nameof(ICurrentEventsByTagQuery), ReadJournal?.GetType().Name ?? "null");
+
+            var allMessages = Enumerable.Range(0, 2000)
+                .Select(i => $"green {i}").ToArray();
+            var a = Sys.ActorOf(TestPersistenceActor.Props("a"));
+            foreach (var str in allMessages)
+            {
+                a.Tell(str);
+                ExpectMsg($"{str}-done");
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Azure.Tests/Query/AzureTableCurrentEventsByTagSpec.cs]: Do a `ReceiveN` outside of the loop instead so we don't have to await each operation individually - will make the test run significantly faster.

File: src/Akka.Persistence.Azure/Akka.Persistence.Azure.csproj [hunk 7]
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <TargetFrameworks>$(NetStandardLibVersion);$(NetVersion)</TargetFrameworks>
     <Description>Akka.Persistence support for Windows Azure Table storage and Azure blob storage.</Description>
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>latest</LangVersion>
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Azure/Akka.Persistence.Azure.csproj:5]: LGTM

File: src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs [hunk 8]
```diff
@@ -285,7 +285,7 @@ protected override bool ReceivePluginInternal(object message)
             {
                 case ReplayTaggedMessages replay:
                     ReplayTaggedMessagesAsync(replay, _shutdownCts.Token)
-                        .PipeTo(replay.ReplyTo, success: h => new RecoverySuccess(h), failure: e => new ReplayMessagesFailure(e));
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs:288]: Ah, is this what was truncating the results stream? Pre-emptively declaring the mission complete?

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Minor suggestion around the test - going to check failed test results now

**Comment by Aaronontheweb** [comment]
What's with the sharding specs failing? Unrelated?

image

File: src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs (hunk 8 above)
Comment (Arkatufus) [reply src/Akka.Persistence.Azure/Journal/AzureTableStorageJournal.cs:288]: Sort of, its only half of the story

**Comment by Arkatufus** [comment]
I'm not sure, let me check

File: src/Akka.Persistence.Azure.Tests/Query/AzureTableCurrentEventsByTagSpec.cs (hunk 6 above)
Comment (Arkatufus) [reply src/Akka.Persistence.Azure.Tests/Query/AzureTableCurrentEventsByTagSpec.cs]: done

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**